### ISSUE TITLE
Blazor (Server) route does not match when complex characters are in the url

### ISSUE THREAD

**Issue by Sebazzz** [dotnet/aspnetcore#53262]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Kind-of a regression of #16840 (fixed in PR #26769), a route is defined for a page with a placeholder. However, when there are weird characters in the URL it doesn't match the route for the page. This worked in .NET 7 and is broken in .NET 8.

In .NET 8 the route does not match:

```
dbug: Microsoft.AspNetCore.Components.Routing.Router[3]
      Navigating to non-component URI 'http://localhost:5110/project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken' in response to path '/project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/'
dbug: Microsoft.AspNetCore.Components.Routing.Router[1]
      Displaying NotFound because path '/project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/' does not match any component route
```

### Expected Behavior

In .NET 7 the route matches and this is logged:

```
dbug: Microsoft.AspNetCore.Components.Routing.Router[2]
      Navigating to component dotnet_bug.Pages.MyComponent in response to path 'project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/'
dbug: Microsoft.AspNetCore.Components.Routing.Router[2]
      Navigating to component dotnet_bug.Pages.MyComponent in response to path 'project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/'
dbug: Microsoft.AspNetCore.Components.Routing.Router[2]
      Navigating to component dotnet_bug.Pages.MyComponent in response to path 'project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/'
dbug: Microsoft.AspNetCore.Components.Routing.Router[1]
      Displaying NotFound because path 'project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/!%40%23%24%25%5E%26*()_%2B-%3D[]%7B%7D//|%3B'%3A/%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/' does not match any component route
dbug: Microsoft.AspNetCore.Components.Routing.Router[1]
      Displaying NotFound because path 'project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/!%40%23%24%25%5E%26*()_%2B-%3D[]%7B%7D//|%3B'%3A/%22%3E%3F.%2F/bewerken' with base URI 'http://localhost:5110/' does not match any component route
```

### Steps To Reproduce

Given a component with this route:

    @page "/project/{projectCode}/ruimte/{publicAreaId}/bron/{sourceCode}/bewerken"

And URL:

    /project/asdasd/ruimte/fb6566f385ab4e11a27f3393bb22c8ff/bron/%21%40%23%24%25%5E%26%2A%28%29_%2B-%3D%5B%5D%7B%7D%5C%5C%7C%3B%27%3A%5C%22%3E%3F.%2F/bewerken

The `sourceCode` part `!@#$%^&*()_+-=[]{}\\|;':\">?./` is url-encoded.

Full repro: [dotnet-bug.zip](https://github.com/dotnet/aspnetcore/files/13884285/dotnet-bug.zip)

### .NET Version

8.0.100

### Anything else?

.NET 7.0:

![net7-0](https://github.com/dotnet/aspnetcore/assets/1426097/8079e80c-3903-4ffd-aced-858cab2e20a5)

.NET 8.0, identical code:

![net8-0](https://github.com/dotnet/aspnetcore/assets/1426097/081f229d-459a-422c-8a76-732ee4d2496d)

**Comment by TDroogers** [comment]
Sounds related to my bug #53138

**Comment by javiercn** [comment]
This is now fixed in main; we'll probably backport it to 8.0 in a future patch release. https://github.com/dotnet/aspnetcore/pull/53470

**Comment by mkArtakMSFT** [comment]
@Sebazzz did you by any chance try the fix out from the nightly build to see if this indeed has addressed the issue you were facing? It most probably did but would be nice to get confirmation.
